jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java

```diff
@@ -421,7 +468,14 @@ public boolean enable(E pooled, boolean acquire)
         @Override
         public E getPooled()
         {
-            return pooled;
+            try
+            {
+                return weakPooled == null ? null : weakPooled.get();
+            }
+            finally
+            {
+                this.pooled = null;
+            }
```

Comment:
This now has a significant side effect of nulling the hard reference.  It is no longer a "getter".  It should be called `takePooled` or `acuirePooled` or similar

Reply:
Should this be an ISE if the pooled has already been taken?

Reply:
Unfortunately, we cannot change this method name as it is part of the API. And I don't see any reason why we should throw ISE.

Reply:
Then is this the right place to null the hard reference?  Should it not be done when the entry is acquired rather than when the item is got from the entry?

Reply:
We cannot null out the hard ref during `acquire()`, since that method returns an `Entry`, there is a time gap between us nulling out the hard ref and the caller of `acquire()` calling `getPooled()` and store the pooled object in a hard ref. During that time, the GC could clear the weak ref.